File: scrapers/torrent_galaxy.js
const express = require("express");
const router = express.Router();
const axios = require("axios");
const cheerio = require("cheerio");
//  Base URL of Torrent Galaxy
const BASE_URL = require("./constants").TORRENTGALAXY_BASE_URL;

router.get("/tgx", async function (req, res) {
    //  Get the Item to be searched from Query
    var search = req.query.search.trim();
    var response = await axios.get(BASE_URL + search).catch((err) => {
        console.log(err);
    });
    var $ = cheerio.load(response.data);
    var jsonResponse = [];
    var selector = $("div.tgxtablerow");
    if (selector.length > 0) {
        selector.each((index, element) => {
            //  File Name
            name = $(element).find("div:nth-child(4)").text();
            //  Seeders
            seeders = $(element).find(
                "div:nth-child(11) span font:nth-child(1)"
            ).text();
            //  Leechers
            leechers = $(element)
                .find("div:nth-child(11) span font:nth-child(2)")
                .text();
            //  Upload Date
            upload_date = $(element).find("div:nth-child(12)").text().split(' ')[0];
            //  File Size
            file_size = $(element).find("div:nth-child(8)").text();
            //  Uploader
            uploader = $(element).find("div:nth-child(7)").text();
            //  magnet
            magnet = $(element)
                .find("div:nth-child(5) a:nth-child(2)")
                .attr("href");
            url =
                "https://torrentgalaxy.to" +
                $(element).find("div:nth-child(4) a").attr("href");

            jsonResponse.push({
                name: name,
                torrent_url: url,
                seeders: seeders,
                leechers: leechers,
                upload_date: upload_date,
                size: file_size,
                uploader: uploader,
                magnet: magnet,
                website: "Torrent Galaxy",
            });
        });
        res.status(200).json({ data: jsonResponse }).end();
    } else {
        res.status(204).end();
    }
});

module.exports = router;

File: routes/Torrentgalaxy.go
package routes

import (
	"fmt"
	"net/url"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/gocolly/colly/v2"
)

func Torrentgalaxy(ginCon *gin.Context) {
	// search := strings.ReplaceAll(strings.TrimSpace(ginCon.Query("search")), " ", "%20")
	param := url.Values{}
	param.Add("search", ginCon.Query("search"))
	url := fmt.Sprintf("https://torrentgalaxy.to/torrents.php?%s", param.Encode())
	c := colly.NewCollector()
	infos := make([]TorrentInfo, 0)
	c.OnHTML("body", func(e *colly.HTMLElement) {
		ti := TorrentInfo{}
		print(e.DOM.Find("div.tgxtablerow").Length())
		if e.DOM.Find("div.tgxtablerow").Length() == 0 {
			return
		}
		e.ForEach("div.tgxtablerow", func(i int, e *colly.HTMLElement) {
			ti.Name = e.ChildText("div:nth-child(4)")
			ti.Seeders = e.ChildText("div:nth-child(11) span font:nth-child(1)")
			ti.Leechers = e.ChildText("div:nth-child(11) span font:nth-child(2)")
			ti.Date = strings.Split(e.ChildText("div:nth-child(12)"), " ")[0]
			ti.Size = e.ChildText("div:nth-child(8)")
			ti.Uploader = e.ChildText("div:nth-child(7)")
			ti.Magnet = e.DOM.Find("#click").Next().Find("a:nth-child(2)").AttrOr("href", "")
			ti.Url = "https://torrentgalaxy.to" + e.ChildAttr("div:nth-child(4) a", "href")
			ti.Website = "Torrent Galaxy"
			ti.TorrentFileUrl = e.DOM.Find("#click").Next().Find("a:nth-child(1)").AttrOr("href", "")
			infos = append(infos, ti)
		})
	})
	c.OnError(func(r *colly.Response, err error) {
		fmt.Println("Request URL:", r.Request.URL, "failed with response:", r, "\nError:", err)
	})
	c.OnScraped(func(r *colly.Response) {
		if len(infos) > 0 {
			ginCon.JSON(200, TorrentRepo{infos})
		} else {
			ginCon.AbortWithStatus(204)
		}
	})
	c.Visit(url)
}
